### ISSUE TITLE
Need unit tests for FT

### ISSUE THREAD

**Issue by mpichbot** [pmodels/mpich#1901]
**Originally by _wbland_ on 2013-07-15 11:01:12 -0500**

---

Now that fault tolerance is coming back in MPICH (see #1889), we need some unit tests to make sure that it works. This ticket will function as a list of all the tests we need. Feel free to add as you see fit. I won't go through and enumerate all of the OS / version combinations that we should use, but generally, I would say that whatever else we test on should be used to test these.
- Checkpoint / Restart (BLCR)
  - Compile with C/R enabled
  - Test that checkpoint can be written successfully
  - Test that checkpoint can be used for restart successfully (I'm fairly certain this isn't working at the moment)
- FT (run with `-disable-auto-cleanup`)
  - Make sure that a regular application still exits correctly (without a failure)
    - Ticket #1537 has an issue with `MPI_Abort`, we should specifically test for that
  - Kill one process and make sure MPICH doesn't crash
    - No communication: just INIT, kill, FINALIZE
    - Using point-to-point communication before/during/after the failure
      - Communicating with dead process (should give error code)
      - Not communicating with dead process (should work fine)
      - Communicating using `MPI_ANY_SOURCE` (should give error code for existing requests when failure is detected and continue to do so for new requests)
    - Check that the failed processes show up in `MPIX_Comm_group_failed`
- FT & Collectives (Once #1889 is closed)
  - Compile with `ENABLE_COLL_FT_RET` on (this will be default once the branch is merged into master)
  - Compile with `ENABLE_COLL_FT_RET` off
  - Kill a processes and make sure collectives propagate the failure correctly
    - Using blocking collectives during the failure (MPICH should be able to fail out of the collectives and return an error)
      - I don't know how many collectives we need to test this on to be representative, but we should at least get some from each type of collective (synchronizing, non-synchronizing, communicator creation, etc.)
    - Using blocking collectives after the failure
    - Using non-blocking collectives during the failure
    - Using non-blocking collectives after the failure
